### ISSUE TITLE
dynamic models with irregular lengths

### ISSUE THREAD

**Issue by karalets** [uber/pyro#34]
Currently we have no very good way of dealing with packed sequences, such as time series of irregular lengths which have been packed into tensors with zero-padding to be able to do minibatch sgd while masking them out.

Pytorch has helper functions for packed sequences, but we need to see how we intend to support such irregularities efficiently without having to process each sample separately.

This may interface weirdly with the mapdata issue I brought up ( https://github.com/uber/pyro/issues/33 ).

Discuss!

**Comment by fritzo** [comment]
I'm also looking into nested `iarange`s with very sparse tensors. I believe PyTorch's sparse tensors could be a viable alternative to masking for very sparse tensors.

**Comment by fritzo** [comment]
Closing since I believe this can now be done by a combination of:
- nested `irange`/`iarange` as of #780 and
- masking via `my_dist.mask(my_mask)` as of #821 
Feel free to re-open if I've misunderstood.

**Comment by PhSteel** [comment]
@fritzo Would love if you could give an example.

**Comment by fritzo** [comment]
@PhSteel here's an example with a batch of time series with different lengths. I haven't tried to run it.
```py
def model(batched_series, lengths):
    mean = pyro.param("mean", torch.tensor(0.0))
    batch_size, max_length = batched_series.shape
    assert lengths.shape == (batch_size,)
    with pyro.iarange('batch', batch_size):
        mask = torch.arange(max_length) < lengths.unsqueeze(-1)
        pyro.sample("series",
                    dist.Normal(mean, 1)
                        .reshape(batched_series.shape)  # expand to the full batch
                        .mask(mask)                     # mask out unwanted terms
                        .reshape(extra_event_dims=1),   # declare time index as dependent
                    obs=batched_series)
```
